clippy_lints/src/needless_move.rs

```diff
@@ -0,0 +1,232 @@
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for closures and `async` blocks where the `move` is not necessary.
+    /// E.g. all the values are captured by value into the closure / `async` block.
+    ///
+    /// ### Why is this bad?
+    /// This pattern is not necessarily bad, but sometimes the `move` keyword is unnecessary,
+    /// for example when there's a closure which captures some variables by reference, so
+    /// the programmer adds the `move` keyword to move the variables into the closure, but
+    /// then later decides that he no longer needs the variables in question, so he removes them
+    /// from the body of the closure, but forgets to also remove the `move` keyword.
+    ///
+    /// This is really just a strict coding style issue.
+    ///
+    /// ### Caveats
+    /// There are some cases where this lint will suggest removing the `move` keyword,
+    /// but it would be considered idiomatic to keep it.
+    ///
+    /// For example, the closure passed to `std::thread::spawn` is usually always written
+    /// with the `move` keyword, even if it's not necessary:
+    ///
+    /// ```no_run
+    /// # fn function_that_does_something_with(_: String) {}
+    /// let a = String::new();
+    /// std::thread::spawn(move || {
+    ///    // ...
+    ///   function_that_does_something_with(a); // a is moved into the closure
+    /// });
+    /// ```
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let a = String::new();
+    /// let closure = move || {
+    ///     drop(a);
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a = String::new();
+    /// let closure = || {
+    ///     drop(a);
+    /// };
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub NEEDLESS_MOVE,
+    restriction,
+    "checks for needless `move`s on closures / `async` blocks"
+}
+
+declare_lint_pass!(NeedlessMove => [NEEDLESS_MOVE]);
+
+impl NeedlessMove {
+    fn check_closure<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, closure: &'tcx Closure<'tcx>) {
+        let CaptureBy::Value { move_kw } = closure.capture_clause else {
+            return;
+        };
+
+        if move_kw.is_dummy() {
+            // async fn ...() {} convert the body to an `async move {}` block,
+            // with a DUMMY_SP for the move_kw
+            return;
+        }
+
+        // Collect moved & borrowed variables from the closure, which the closure *actually* needs.
+        let ctx = {
+            let mut ctx = MovedVariablesCtxt::default();
+            let body = cx.tcx.hir().body(closure.body);
+            let infcx = cx.tcx.infer_ctxt().build();
+            euv::ExprUseVisitor::new(&mut ctx, &infcx, closure.def_id, cx.param_env, cx.typeck_results())
+                .consume_body(body);
+            ctx
+        };
+
+        let mut lint_result = LintResult::NothingCaptured;
+
+        for captured_place in cx.typeck_results().closure_min_captures_flattened(closure.def_id) {
```

Comment:
Can you run me down on this? What difference does using `ExprUseVisitor` make to just calling this, and checking if any are `UpvarCapture::ByRef(_)`? Is it because we need to filter out the captured variables which aren't used (which would somehow show up in `closure_min_captures_flattened`)?

Reply:
Ah, nvm, I assume this is relevant to https://github.com/rust-lang/rust-clippy/pull/11759#discussion_r1385165647?